Feature/#6603 apportionment process for campd

djw4erg opened the pull request:
## Overview

Created database objects to support the generation of apportioned emissions.

## Tables

- PDEM_Report: the parent table indicating emission reports that need to be processed for a submission.  Also used to queue and track the job that generates the apportioned emissions.
- PDEM_MATS_Monitor_Hour: Contains the MATS monitor hourly data used to apportion to the unit level.
- PDEM_MATS_Unit_Hour: Contains the apportioned MATS unit hourly data and used to populate the CAMDDMW/CAMDDMW_ARCH emissions data.
- PDEM_P75_Monitor_Hour: Contains the Part 75 monitor hourly data used to apportion to the unit level.
- PDEM_P75_Unit_Hour: Contains the apportioned Part 75 unit hourly data and used to populate the CAMDDMW/CAMDDMW_ARCH emissions data.
- PDEM_Report_User: Tracks which 'users' of the apportioned emissions data have used the data for the submission.

## Stored Procedures and Functions

### Apportionment Formula Data Table Routines

Return the data used to apply custom apportionment formulas for complex and changing configurations.

- pdem_apportionment_get_data_table
- pdem_apportionment_get_formula_table
- pdem_apportionment_get_range_table
- pdem_apportionment_get_subtractive_table
- pdem_apportionment_get_table

### Update Stored Routines

- pdem_delete_unit_data
- pdem_update_failure
- pdem_update_init
- pdem_update_init_get_active_link_counts
- pdem_update_init_get_combined_hourly_data
- pdem_update_init_get_cp_ltff_info
- pdem_update_init_get_location_info
- pdem_update_init_get_location_type_counts
- pdem_update_init_get_report_dates
- pdem_update_init_get_reporting_period_info
- pdem_update_init_get_spanning_link_counts
- pdem_update_init_load_mats_monitor_hour
- pdem_update_init_load_p75_monitor_hour
- pdem_update_init_reset
- pdem_update_success

### Helper Routines

- pdem_is_mats_emission_report
- pdem_mats_modc_to_measure_code

### Update CAMDDMW and CAMDDMW_ARCH Routines

- pdem_update_public
- pdem_update_public_load_mats_unit_hour
- pdem_update_public_load_p75_unit_annual
- pdem_update_public_load_p75_unit_day
- pdem_update_public_load_p75_unit_hour
- pdem_update_public_load_p75_unit_month
- pdem_update_public_load_p75_unit_ozone
- pdem_update_public_load_p75_unit_quarter

### Job Enqueue and Dequeue Routines

- pdem_job_get_next
- pdem_job_queue

djw4erg committed:
#3952 Addition of core Program Data Emission (PDEM) tables.

djw4erg committed:
#3952 Addition of PDEM_REPORT_USER and PDEM_REPORT_USER_CODE table DDLs.

djw4erg committed:
#3952 Batch of Routines

djw4erg committed:
#3952 Update Init implementation.

djw4erg committed:
#3952 Included Is MATS Emission Report checking.

djw4erg committed:
#3952 Additional Update Init table functions and related stored procedure

djw4erg committed:
#3952 Renamed Apportionment scripts/functions.

djw4erg committed:
#3952: Completed Success and Failure actions including adding emissions_created_flg to PDEM_REPORT table and table updates, mostly to eliminate Userid and Document_Source.  Also corrected exception handling for PDEM_Update_Init.

djw4erg committed:
#6603  Changes in prep for queuing the job, supporting multiple queuing of the same emission report, and as a result of testing with the .Net code.

djw4erg committed:
#6603 Renaming files.

djw4erg committed:
#6603

1. Change to update the "public" tables in CAMDDMW and CAMDDMW_ARCH, including the aggregate tables.
2. Corrected missing Pollutant_Cd  from PDEM_Apportionment_Get_Formula_Table.
3. Made loading of P75 and MATS CAMDECMPSAUX hourly data separate from the PDEM_Update_Success SP.
4. Replaced summing of aggregate data in CAMDECMPSAUX with updating of P75 and MATS CAMDDMW (and CAMDDMW_ARCH) hourly data and summing of aggregate P75 data.

djw4erg committed:
#6603 Addition of Triggered Time column, Status Code calculated column, and check constraints controlling the setting of times.

djw4erg committed:
#6603 Implementing PDEM queuing and get next stored procedure and function changes.

djw4erg committed:
Merge pull request #114 from US-EPA-CAMD/develop

feature/#6603-Apportionment-Update-From-Develop

djw4erg committed:
index on feature/#6603-Apportionment-Process-for-CAMPD: 9746be8 #6603 Implementing PDEM queuing and get next stored procedure and function changes.

djw4erg committed:
On feature/#6603-Apportionment-Process-for-CAMPD: 2025-03-18

djw4erg committed:
Merge commit 'e5ea1f0774a94f46713fbe47b81a6b5b5ddd69b8' into feature/#6603-Apportionment-Process-for-CAMPD

djw4erg committed:
#6603 Addition of scripts to deployment folder.

djw4erg committed:
Merge pull request #119 from US-EPA-CAMD/develop

Updating branch from Develop

djw4erg committed:
#6603 Addition of job_configuration row for PDEM.

yonatan-dp commented on camdecmpsaux/constraints-indexes/1-pdem_report.sql:
```diff
@@ -0,0 +1,15 @@
+	ADD CONSTRAINT pdem_report_prd_fk FOREIGN KEY ( rpt_period_id ) REFERENCES camdecmpsmd.reporting_period( rpt_period_id ),
+	ADD CONSTRAINT pdem_report_sub_fk FOREIGN KEY ( submission_id ) REFERENCES camdecmpsaux.submission_queue( submission_id ),
+    ADD CONSTRAINT pdem_report_stm_ck CHECK ( triggered_time is not null or started_time is null ),
+    ADD CONSTRAINT pdem_report_ctm_ck CHECK ( triggered_time is not null and started_time is not null and note_time is null or completed_time is null ),
```
This is minor but I am wondering if we can additional parentheses for clarity?  Not sure by just reading it if the last two conditions are on their own joined by the and...   Saw these in a few places...

djw4erg replied:
Yonatan, the indentation difference is not intentional.  It looks like its the difference between a tab character and spaces for a tab.  I like spaces and switched the IDE half way through working on this.  I am not sure why the PK has spaces, but I definitely added the check constraints afterward.  I will correct and upload an adjusted script.

djw4erg replied:
Yonatan, the indentation difference is not intentional and the separate constraints should have the same indentation.  It looks like its the difference between a tab character and spaces for a tab.  I will update the script.

yonatan-dp commented on camdecmpsaux/tables/pdem_report.sql:
```diff
@@ -0,0 +1,46 @@
+COMMENT ON COLUMN camdecmpsaux.pdem_report.apportionment_type_cd IS 'Code used to indicate the apportionment type of an emissions report.';
+COMMENT ON COLUMN camdecmpsaux.pdem_report.queued_time IS 'The timestamp for when the generation of the Program Data Emissions was gueued.';
+COMMENT ON COLUMN camdecmpsaux.pdem_report.started_time IS 'The timestamp for when the generation of the Program Data Emissions was triggered.';
+COMMENT ON COLUMN camdecmpsaux.pdem_report.started_time IS 'The timestamp for when the generation of the Program Data Emissions was started.';
```
Duplicate comment for pdem_report.started_time? or is one for triggered_time?

djw4erg replied:
One should be for Triggered Time.

djw4erg committed:
#6603 Corrected indentation problem and incorrect column name for column comment.

yonatan-dp commented on camdecmpsaux/procedures/pdem_update_public_load_p75_unit_day.sql:
```diff
@@ -0,0 +1,272 @@
+as $procedure$
+
+declare
+    cRoutineName constant varchar := 'PDEM_Update_Public_Load_Unit_Day';
```
Typo?  should the value be `PDEM_Update_Public_Load_P75_Unit_Day`. I saw this across the *p75* names sql files.

djw4erg replied:
Corrected the following:

- pdem_update_public_load_p75_unit_annual.sql
- pdem_update_public_load_p75_unit_day.sql
- pdem_update_public_load_p75_unit_hour.sql
- pdem_update_public_load_p75_unit_month.sql
- pdem_update_public_load_p75_unit_ozone.sql
- pdem_update_public_load_p75_unit_quarter.sql

yonatan-dp commented on camdecmpsaux/tables/pdem_mats_monitor_hour.sql:
```diff
@@ -0,0 +1,64 @@
+    hit                     numeric(15,3),
+    hit_hour_measure_cd     varchar(7),
+	hit_From_Fuel_Flow      numeric(14, 4) NULL,
+    hg_rate_eo              varchar(35),
```
Should this be numeric?  I did a global search for `hg_rate_eo` in db-scripts and I see references as numeric. I checked `cHourlyRawData.cs` (line 626 in your branch) in quartz-scheduler and I see this: `row["HG_RATE_EO"].AsDecimal(),`.  
Don't know if I am looking at something different. I see `hg_rate_hi` and `hcl_rate_eo` defined as decimal too...

djw4erg replied:
This process initially pulls the MATS values from the camdecmps MATS_DERIVED_HRLY_VALUE table that stores the various RATE_EO and RATE_HI values in scientific notation as varchar(30).  The camdecmpsaux PDEM_MATS_MONITOR_HOUR table continues using scientific notation since pulls directly from MATS_DERIVED_HRLY_VALUE.  The processing in the EASEY-QUARTZ-SCHEDULER Check Engine solution converts the values from scientific notation to decimals before populating the camdecmpsaux PDEM_MATS_UNIT_HOUR table which is used to populate the camddmw HOUR_UNIT_MATS_DATA table.

The RATE_EO and RATE_HI columns should be varchar in PDEM_MATS_MONITOR_HOUR and should be decimal in PDEM_MATS_UNIT_HOUR and HOUR_UNIT_MATS_DATA (and its related log and archive tables).

yonatan-dp commented on misc/load-job-configurations.sql:
```diff
@@ -12,6 +12,7 @@ INSERT INTO camdaux.job_configuration (job_type, job_group, job_name, job_descri
     ('EvaluationJobQueue', 'MAINTAINANCE', 'Evaluation Job Queue', 'Operates on an interval to determine if files in evaluation queue can be triggered.', 'Check evaluation queue every minute', 'Operate every minute to determine if there are files in evaluation queue which can be triggered', '0 0/1 * 1/1 * ? *', TRUE, FALSE, NULL),
```
Has this branch been updated from develop already?  This line is different in develop (I made a change to run this job every 15 seconds last sprint).  Ignore if this branch does not have the latest from develop. If it does, then merge error?

djw4erg replied:
This branch is behind by a number of commits including https://github.com/US-EPA-CAMD/easey-db-scripts/commit/b04745ee96073903e23ad6750bb9f35ef6ca148a.

djw4erg committed:
#6603 PR Review Corrections

1) Corrected cRoutineName constant in PDEM_UPDATE_PUBLIC_LOAD_P75_UNIT SP scripts.
2) Removed no longer existing tables from PDEM_UPDATE_INIT_RESET SP script.
